### PR TITLE
Training Loop and Baseline Model

### PR DESCRIPTION
**Pull request by gserifi**
Tons of changes, so just look at the final version.

### CODE REVIEW

File: src/blocks/ff.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        )
+
+        self.ff = nn.Sequential(
+            proj,
+            nn.Dropout(dropout),
+            nn.Linear(hidden_dim, dim_out),
+        )
+
+    def forward(self, x: Tensor) -> Tensor:
+        return self.ff(x)
```
Comment (ucinereo) [inline src/blocks/ff.py:83]: Not focus this project, but might still be nice to make the activation configurable and just default to GEGLU. What do  you think?

File: src/blocks/mha.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    def forward(
+        self,
+        q: Tensor,
+        ctx: Tensor,
+        mask: torch.Tensor = None,
+    ) -> tuple[Tensor, Tensor]:
+        q, k, v = map(lambda f, x: f(x), (self.query, self.key, self.value), (q, ctx, ctx))
+        q, k, v = map(lambda x: rearrange(x, "b s (h d) -> b h s d", h=self.n_heads), (q, k, v))
+
+        scores = torch.einsum("b h s d, b h t d -> b h s t", q, k) * self.scale
```
Comment (ucinereo) [inline src/blocks/mha.py:61]: With each `einsum` expression, I would consider writing a short comment what it does, for people who are not so familiar with `einsum`. Although this here should be straight forward :)


```suggestion
        # Q @ K.T
        scores = torch.einsum("b h s d, b h t d -> b h s t", q, k) * self.scale
```

File: src/blocks/mha.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+        if mask is not None:
+            # Expand mask to match scores shape: (b, s, t) -> (b, 1, s, t)
+            mask = mask.unsqueeze(1)
+            scores = scores.masked_fill(mask == 1, float("-inf"))
+
+        attn = scores.softmax(dim=-1)
+
+        out = torch.einsum("b h s t, b h t d -> b h s d", attn, v)
```
Comment (ucinereo) [inline src/blocks/mha.py:72]:
```suggestion
        # A @ V
        out = torch.einsum("b h s t, b h t d -> b h s d", attn, v)
```

File: src/blocks/ff.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                nn.Linear(dim, hidden_dim),
+                nn.GELU(),
+            )
+        )
+
+        self.ff = nn.Sequential(
+            proj,
+            nn.Dropout(dropout),
+            nn.Linear(hidden_dim, dim_out),
+        )
```
Comment (ucinereo) [inline src/blocks/ff.py:80]: Not focus of this project, but I think it would still be nicer from an abstraction point of view to make the activation function configurable. What do you think?

File: src/blocks/mha.py (hunk 2 above)
Comment (ucinereo) [inline src/blocks/mha.py:61]: For `einsum` expressions, I would add a simple comment about what it does, for people who are not familiar with it. Although this one here should be fairly simple to understand.


```suggestion
        # Q @ K.T
        scores = torch.einsum("b h s d, b h t d -> b h s t", q, k) * self.scale
```

File: src/blocks/mha.py (hunk 3 above)
Comment (ucinereo) [inline src/blocks/mha.py:72]:
```suggestion
        # A @ V
        out = torch.einsum("b h s t, b h t d -> b h s d", attn, v)
```

File: src/datasets/addition.py [hunk 5]
```diff
@@ -1,56 +1,103 @@
...
+            seed: Random seed for reproducibility.
+        """
+        self.n_samples = n_samples
         self.max_digits = max_digits
+        self.h = 5
+        self.w = max_digits + 3 + 1
         self.seq_len = 2 * (max_digits + 1)
         self.seed = seed
         self.data = self._generate_data()
         self.token_to_idx = list("0123456789+ -_\n")
```
Comment (ucinereo) [inline src/datasets/addition.py:29]: Why do you consider `\n` a token? IMO this should be hidden from the set as it's implicitly stored in the tensor dimension. The addition algorithm should not be thought of as a string, where newlines are necessary. This will be a distinction from the 1D chain of thought, no?

File: notebooks/analysis.ipynb
Comment (ucinereo) [inline notebooks/analysis.ipynb:1]: Put into folder `notebooks` or completely leave out.

File: train.py [hunk 6]
```diff
@@ -1,16 +1,300 @@
...
+from tqdm import tqdm
+
+from src.interfaces import DatasetBase, TransformerBase, TransformerHeadBase
+from src.utils import build_data_loader, drop_helpers
+
+logger = logging.getLogger(__name__)
+
+
+class Trainer:
+    def __init__(self, cfg: DictConfig, output_dir: Path):
```
Comment (ucinereo) [inline train.py:25]: Just to follow [SOLID](https://en.wikipedia.org/wiki/SOLID), maybe split up into different `_init_logging`, `_init_data`, etc.

File: src/models/transformer.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        src_seq = rearrange(src, "b h w -> b (h w)")
+        tgt_seq = rearrange(tgt, "b h w -> b (h w)")
+
+        # Embed tokens: (b, s) -> (b, s, d)
+        src_emb = self.src_embed(src_seq)
+        tgt_emb = self.tgt_embed(tgt_seq)
+
+        # Add positional encoding
+        src_enc = self.pos_encoding(src_emb, src_orig_size)
+        tgt_enc = self.pos_encoding(tgt_emb, tgt_orig_size)
```
Comment (ucinereo) [inline src/models/transformer.py:85]: Might need to abstract this further in the future if we want different pos encodings like the rotary one in eope.

File: train.py [hunk 8]
```diff
@@ -1,16 +1,300 @@
...
+        self.log_writer.add_scalar("train/avg_grad_norm", avg_grad_norm, step)
+
+    def forward(
+        self, batch: tuple[Tensor, Optional[Tensor]]
+    ) -> tuple[
+        Tensor, Tensor, Tensor, Optional[Tensor], Optional[Tensor], Optional[Tensor], Optional[Tensor], Optional[Tensor]
+    ]:
+        x_in, x_out = batch
+
+        src, tgt = self.model.prepare_tokens(x_in, x_in, self.head)  # todo: check if this makes sense
```
Comment (ucinereo) [inline train.py:151]: shouldn't the second parameter be `x_out`?

**Review by ucinereo — COMMENTED** [review]
I've added some suggestions and comments about things which are currently unclear to me. Very nice PR!

File: src/blocks/ff.py (hunk 1 above)
Comment (gserifi) [reply src/blocks/ff.py:83]: Makes sense, done

File: src/models/transformer.py (hunk 7 above)
Comment (gserifi) [reply src/models/transformer.py:85]: Since 1D sinusoidal and RoPE are applied at different locations, I was thinking that one can just use nn.Identity for pos_encoding when using RoPE and then set some flag in the MHA. Let's discuss this later.

File: train.py (hunk 8 above)
Comment (gserifi) [reply train.py:151]: Currently, that doesn't matter because we focus on encoder-only. I think one could either pass x_in, or x_out if you additionally pass an attention mask, if using decoder.

File: train.py (hunk 6 above)
Comment (gserifi) [reply train.py:25]: Good idea

File: src/datasets/addition.py (hunk 5 above)
Comment (gserifi) [reply src/datasets/addition.py:29]: Honestly, this is a big question we should clarify asap.
